### PR TITLE
RAINCATCH-911 - Sync cloud backbone

### PR DESCRIPTION
**Pull request by wtrocki**
h4. Motivation

Create sync wrapper for cloud.

h4. Status

- [x] backbone
- [x] custom handlers setup
- [x] unit tests
- [x] functional tests
- [x] documentation (separate repo)
- [x] readme

### CODE REVIEW

**Review by austincunningham — COMMENTED** [review]
👍

File: cloud/datasync/README.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    // Create api object
+    const sync: SyncApi = new SyncServer();
+    // Options for sync connection
+    const connectOptions: SyncOptions = {},
+    // Connect database and start sync
+    sync.connect(connectOptions, function (err) {
+      if (err) {
+        console.log(err);
+      }
+    })
```
Comment (paolobueno) [inline cloud/datasync/README.md]: no need for the 4 spaces when using ```, this makes the code indented when the markdown's rendered

File: cloud/datasync/README.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+```typescript
+//middleware
+app.use(bodyParser.json());
+app.use(cors());
+
+// Use api path of your wish
+// Same endpoint needs to be configured in js client
+const path = '/sync/:datasetId';
+const middleware: SyncExpressMiddleWare = new
+SyncExpressMiddleWare();
```
Comment (paolobueno) [inline cloud/datasync/README.md]: random line break here

File: cloud/datasync/README.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+const router = middleware.createSyncExpressRouter();
+app.use(router);
```
Comment (paolobueno) [inline cloud/datasync/README.md]: maybe you meant `app.use(path, router)` ?

File: cloud/datasync/README.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+```
+
+2. Create api endpoint for sync.
+
+This endpoint will be called by all clients that wish to synchronize their local data with server.
+
+```typescript
+//middleware
+app.use(bodyParser.json());
+app.use(cors());
```
Comment (paolobueno) [inline cloud/datasync/README.md:42]: Are these middleware required for sync to work? If not I think it's best not to include them in the examples

File: cloud/datasync/README.md (hunk 3 above)
Comment (wtrocki) [reply cloud/datasync/README.md]: Router itself accept path. This works fine, but can replace with `app.use("/"..`

File: cloud/datasync/README.md (hunk 4 above)
Comment (wtrocki) [reply cloud/datasync/README.md:42]: Actually they are essential to be there, but we did not want them to by part of the middleware as this stuff is usually configured for root.

What we saying here is that when you have sync you need to have json api capabilities and cors() securtity

File: cloud/datasync/src/SyncServer.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+import { setupGlobalHandlers, setupHandlers } from './crud-handlers/HandlerMapper'
+import DataSetHandler from './crud-handlers/DataSetHandler'
+
+import { api } from 'fh-sync'
+
+/**
+ * Implementation for sync server side api
+ */
+export class SyncServer implements SyncApi {
```
Comment (paolobueno) [inline cloud/datasync/src/SyncServer.ts]: This class seems to be mostly calling into fh-sync methods, which are singletons from the way `require/import` works, so wouldn't you have some issues with multiple instances of it?

You don't need to have a constructor function/class to implement an API, specially if you don't have instance members and so on. Just an object is okay.

File: cloud/datasync/README.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+```typescript
+class MySQLDataSetHandler implements DataSetHandler {
+  // We can implement any any method we wish
+  onList(datasetId: string, query: any, metaData: any, callback: (err: Error | string | undefined, res: any | undefined) => void) {
+    let taskList = mysqlTaskRepository.getTasks(query);
+    callback(undefined, taskList);
+  }
+}
+
+sync.registerDatasetDataHandler("task", options, new MySQLDataSetHandler());
```
Comment (paolobueno) [inline cloud/datasync/README.md]: No need for a class to implement this too, could be just an object, even an inline one with  `{ onList: function(...){...} }`

Comment (paolobueno) [reply cloud/datasync/README.md]: an inline object should also trigger editor autocomplete

Comment (wtrocki) [reply cloud/datasync/README.md]: I had this before actually and I think that actually for the purpose of getting started class here is more clear for reader.

File: cloud/datasync/src/SyncServer.ts (hunk 5 above)
Comment (wtrocki) [reply cloud/datasync/src/SyncServer.ts]: How about our extensibility story here then?

**Comment by wtrocki** [comment]
@paolobueno Applied all comments

File: cloud/datasync/README.md (hunk 4 above)
Comment (paolobueno) [reply cloud/datasync/README.md:42]: :+1: that's what I thought, if they're required by sync then leave them there :D

**Comment by wtrocki** [comment]
@paolobueno This PR is blocked by build issues ticket I mentioned.

File: README.md [hunk 7]
```diff
@@ -36,6 +36,9 @@ contained in the following directories:
 
   <dt>examples/</dt>
   <dd>Smaller examples and templates for other packages in the repository</dd>
+
+   <dt>integration/</dt>
+  <dd>Examples how Raincatcher modules integrate with each other in form of functional tests</dd>
```
Comment (JameelB) [inline README.md]: Examples `of` how Raincatcher modules integrate with each other in `the` form of functional tests

File: cloud/datasync/README.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
+# RainCatcher Sync Cloud
+
+Raincatcher Wrapper for Feedhenry DataSync server
```
Comment (JameelB) [inline cloud/datasync/README.md]: Forgot `.` at end of the line.

File: cloud/datasync/README.md [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+This module creates server side service that allows
+mobile applications to sync offline data to preferred database. By default server connects and stores all client data in mongodb database. This behavior can be changed by providing custom handlers.
+
+For advanced usage see FeedHenry sync documentation:
+https://github.com/feedhenry/fh-sync/docs
```
Comment (JameelB) [inline cloud/datasync/README.md]: I'm getting a 404 on this link?

File: cloud/datasync/README.md [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+> **Note**: Cloud service is utilized by one of the sync clients. For more information about client go to: https://github.com/feedhenry/fh-sync-js
+
+## Quick start
+
+1. Start sync service process
+Starting sync service that will monitor and process
+new tasks that are sent from sync clients
```
Comment (JameelB) [inline cloud/datasync/README.md]: The title and description are all in one line.

File: cloud/datasync/README.md [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+For advanced usage see FeedHenry sync documentation:
+https://github.com/feedhenry/fh-sync/docs
+
+> **Note**: Cloud service is utilized by one of the sync clients. For more information about client go to: https://github.com/feedhenry/fh-sync-js
+
+## Quick start
+
+1. Start sync service process
+Starting sync service that will monitor and process
```
Comment (JameelB) [inline cloud/datasync/README.md]: remove `that`

File: cloud/datasync/README.md [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+new tasks that are sent from sync clients
+
+```typescript
+import SyncServer, { SyncApi, SyncOptions, SyncExpressMiddleWare } from '../src/index'
+
+// Assign default implementation to sync interface
+const sync: SyncApi = SyncServer;
+// Options for sync connection
+const connectOptions: SyncOptions = {},
```
Comment (JameelB) [inline cloud/datasync/README.md:26]: are the available SyncOptions described in the fh-sync docs?

File: cloud/datasync/README.md [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+## Custom data handlers
+
+By default sync will use the same mongodb connection for retrieving and storing any client data. Sync allows to modify this behavior by providing custom `DataSetHandler`
+
+1. Create `DataSetHandler`
+
+```typescript
+class MySQLDataSetHandler implements DataSetHandler {
+  // We can implement any any method we wish
```
Comment (JameelB) [inline cloud/datasync/README.md]: double `any`

File: cloud/datasync/README.md [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+  onList(datasetId: string, query: any, metaData: any, callback: (err: Error | string | undefined, res: any | undefined) => void) {
+    let taskList = mysqlTaskRepository.getTasks(query);
+    callback(undefined, taskList);
+  }
+}
+
+sync.registerDatasetDataHandler("task", options, new MySQLDataSetHandler());
```
Comment (JameelB) [inline cloud/datasync/README.md]: is this line supposed to be under `2. Add handler`?

File: cloud/datasync/README.md [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+
+```typescript
+import { NativeSync } from '../src/index'
+NativeSync.api.anyapicall ;
+```
+
+Where `anyapicall` is one of the types defined in:
+https://github.com/feedhenry/fh-sync/blob/master/types/fh-sync.d.ts
+
+For more advanced usages please follow documentation: https://github.com/feedhenry/fh-sync/docs
```
Comment (JameelB) [inline cloud/datasync/README.md]: getting a 404 not found on this link.

**Review by witmicko — APPROVED** [review]
lgtm

File: cloud/datasync/src/crud-handlers/HandlerMapper.ts [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+import SyncDataSetOptions from '../options/SyncDatasetOptions';
+import DataSetHandler from '../crud-handlers/DataSetHandler'
+
+import * as sync from 'fh-sync'
+
+
+/**
+ * Maps dataHandler to sync api methods
+ *
+ * @param syncAPI
```
Comment (JameelB) [inline cloud/datasync/src/crud-handlers/HandlerMapper.ts]: wrong param description?

File: cloud/datasync/src/crud-handlers/HandlerMapper.ts [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+  }
+  if (dataHandler.onDelete) {
+    sync.handleDelete(datasetId, dataHandler.onDelete)
+  }
+}
+
+/**
+ * Maps global dataHandler to sync api methods
+ *
+ * @param syncAPI
```
Comment (JameelB) [inline cloud/datasync/src/crud-handlers/HandlerMapper.ts]: no syncAPI param here?

File: cloud/datasync/README.md (hunk 12 above)
Comment (wtrocki) [reply cloud/datasync/README.md:26]: They are self documenting (any IDE will provide intelisense and documentation)

**Review by paolobueno — APPROVED** [review]

**Pull request merged**